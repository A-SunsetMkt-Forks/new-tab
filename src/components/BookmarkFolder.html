<div
  class="item folder{endNode ? ' right' : ''}"
  title="{_node.title}"
  on:mouseenter="handleMouseEnter()"
  on:mouseleave="handleMouseLeave()"
>
  {shorten(_node.title, maxLen)}

  {#if lvl !== 0}
    <div class="caret">â–¸</div>
  {/if}

  {#if isOpen}
    <div class="subfolder{endNode ? ' left' : ''}">
      {#each _node.children as childNode}
        <BookmarkNode _node="{childNode}" lvl="{lvl + 1}" maxLen="{40}" />
      {:else}
        <div class="item">(empty)</div>
      {/each}
    </div>
  {/if}
</div>

<script>
  import BookmarkNode from './BookmarkNode.js';
  import { shorten } from '../common.js';

  const EVENT_NAME = 'open';
  const OPEN_DELAY = 200;
  const CLOSE_DELAY = 400;

  export default {
    helpers: {
      shorten,
    },
    components: {
      BookmarkNode,
    },
    // data: () => ({
    //   isOpen: false,
    //   endNode: false,
    // }),
    oncreate() {
      this.openTimer = null;
      this.closeTimer = null;
    },
    ondestroy() {
      // clean up listener to prevent memory leaks
      if (this.openListener) this.openListener.cancel();
    },
    methods: {
      // set up event listener to close subfolder when another opens
      onOpen() {
        this.openListener = this.root.on(EVENT_NAME, (lvl) => {
          if (this.get().isOpen) {
            if (lvl === this.get().lvl) {
              this.set({ isOpen: false });
              this.resetCloseTimer();
            }
          }
        });
      },

      // subfolder mouse enter
      handleMouseEnter() {
        if (this.get().isOpen) {
          this.resetCloseTimer();
        } else {
          // delay to prevent accidental folder open
          this.openTimer = setTimeout(() => {
            this.root.fire(EVENT_NAME, this.get().lvl);
            this.set({ isOpen: true });
            this.onOpen();
          }, OPEN_DELAY);
        }
      },

      // subfolder mouse leave
      handleMouseLeave() {
        // reset open timer
        if (this.openTimer !== null) {
          clearTimeout(this.openTimer);
          this.openTimer = null;
        }

        // set up subfolder close timer
        if (this.get().isOpen) {
          this.closeTimer = setTimeout(() => {
            this.set({ isOpen: false });
          }, CLOSE_DELAY);
        }
      },

      resetCloseTimer() {
        if (this.closeTimer !== null) {
          clearTimeout(this.closeTimer);
          this.closeTimer = null;
        }
      },
    },
  };
</script>

<style type="text/postcss" global>
  .folder {
    position: relative;
  }

  .caret {
    padding: 0 6px;
    margin-left: auto;
  }

  :global(.subfolder) {
    position: absolute;
    top: 40px;
    left: 0;
    background: var(--c0);
    box-shadow: var(--s0);

    & & {
      top: 0;
      left: 100%;
    }
  }

  /* bookmark item pushed to the right */
  .right {
    width: 100%;
    justify-content: flex-end;
  }

  /* folder which opens to the left */
  .left {
    right: 0;
    left: initial;
    justify-content: initial;

    :global(.subfolder) {
      right: 100%;
      left: initial;
    }
  }
</style>
