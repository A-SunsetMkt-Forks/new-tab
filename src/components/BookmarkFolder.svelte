<script>
  /* eslint-disable no-underscore-dangle */

  import { onDestroy } from 'svelte';
  import BookmarkNode from './BookmarkNode'; // eslint-disable-line import/no-cycle
  import { shorten } from '../common';

  // props
  export let lvl = 0;
  export let maxLen = 0;
  export let endNode = false;
  export let _node;

  // reactive data
  let isOpen = false;

  const EVENT_NAME = 'open';
  const OPEN_DELAY = 200;
  const CLOSE_DELAY = 400;

  const openEvent = new CustomEvent(EVENT_NAME, { detail: lvl });
  let openTimer = null;
  let closeTimer = null;

  function resetCloseTimer() {
    if (closeTimer !== null) {
      clearTimeout(closeTimer);
      closeTimer = null;
    }
  }

  /**
   * Close subfolder when another opens.
   * @param {CustomEvent} event Event dispatched when another subfolder opens.
   */
  function handleOpen(event) {
    if (isOpen && event.detail === lvl) {
      isOpen = false;
      resetCloseTimer();
    }
  }

  function onOpen() {
    window.dispatchEvent(openEvent);
    window.addEventListener(EVENT_NAME, handleOpen);
    isOpen = true;
  }

  // subfolder mouse enter
  function handleMouseEnter() {
    if (isOpen) {
      resetCloseTimer();
    } else {
      // delay to prevent accidental folder open
      openTimer = setTimeout(onOpen, OPEN_DELAY);
    }
  }

  // subfolder mouse leave
  function handleMouseLeave() {
    // reset open timer
    if (openTimer !== null) {
      clearTimeout(openTimer);
      openTimer = null;
    }

    // set up subfolder close timer
    if (isOpen) {
      closeTimer = setTimeout(() => {
        isOpen = false;
        window.removeEventListener(EVENT_NAME, handleOpen);
      }, CLOSE_DELAY);
    }
  }

  onDestroy(() => {
    window.removeEventListener(EVENT_NAME, handleOpen);
  });
</script>

<style type="text/postcss">
  :global(.folder) {
    position: relative;
  }

  :global(.caret) {
    padding: 0 6px;
    margin-left: auto;
  }

  :global(.subfolder) {
    position: absolute;
    top: 40px;
    left: 0;
    background: var(--c0);
    box-shadow: var(--s);

    & & {
      top: 0;
      left: 100%;
    }
  }

  /* bookmark item pushed to the right */
  :global(.right) { /* stylelint-disable-line no-descending-specificity */
    justify-content: flex-end;
    width: 100%;
  }

  /* folder which opens to the left */
  :global(.left) { /* stylelint-disable-line no-descending-specificity */
    right: 0;
    left: initial;
    justify-content: initial;

    :global(.subfolder) {
      right: 100%;
      left: initial;
    }
  }
</style>

<div
  class="item folder{endNode ? ' right' : ''}"
  title="{_node.title}"
  on:mouseenter="{handleMouseEnter}"
  on:mouseleave="{handleMouseLeave}"
>
  {shorten(_node.title, maxLen)}

  {#if lvl !== 0}
    <div class="caret">â–¸</div>
  {/if}

  {#if isOpen}
    <div class="subfolder{endNode ? ' left' : ''}">
      {#each _node.children as childNode}
        <BookmarkNode _node="{childNode}" lvl="{lvl + 1}" maxLen="{40}" />
      {:else}
        <div class="item">(empty)</div>
      {/each}
    </div>
  {/if}
</div>
