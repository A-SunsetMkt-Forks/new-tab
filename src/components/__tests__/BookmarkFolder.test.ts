import BookmarkFolder from '../BookmarkFolder.svelte';

const node = {
  title: 'Test Bookmark',
};

describe('BookmarkFolder component', () => {
  it('renders correctly', () => {
    expect.assertions(1);
    const target = document.createElement('div');
    new BookmarkFolder({
      props: { node },
      target,
    });
    expect(target.innerHTML).toMatchSnapshot();
  });
});
