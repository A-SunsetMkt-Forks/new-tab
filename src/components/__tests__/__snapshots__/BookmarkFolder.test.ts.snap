// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookmarkFolder component renders correctly 1`] = `
"<div class=\\"item folder \\" title=\\"Test Bookmark\\">Test Bookmark

  

  </div>"
`;
